### DISCUSSION TITLE
Incomplete Dispose of WatsonTCPServer

### DISCUSSION THREAD

**Discussion by sancheolz** [jchristn/WatsonTcp#331] [Q&A]
This test fails starting from 6.0 version of WatsonTCP:
```cs
var server = new WatsonTcpServer("127.0.0.1", 9000);
server.Events.MessageReceived += (_, _) => { Console.WriteLine("srecv"); };
server.Start();
var client = new WatsonTcpClient("127.0.0.1", 9000);
client.Events.MessageReceived += (_, _) => { Console.WriteLine("crecv"); };
client.Connect();
Task.Run(() => { client.SendAsync("aaa"); });
server.Dispose();
```
It throws
```
System.AggregateException : One or more errors occurred. (A task was canceled.)
  ----> System.Threading.Tasks.TaskCanceledException : A task was canceled.
   at System.Threading.Tasks.Task.ThrowIfExceptional(Boolean includeTaskCanceledExceptions)
   at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at System.Threading.Tasks.Task.Wait()
   at WatsonTcp.WatsonTcpServer.Dispose(Boolean disposing) in WatsonTcpServer.cs:line 621
```

**Comment by ShayanFiroozi** [comment]
Hi , 

**`Task.Run(() => { client.SendAsync("aaa"); });`**

You are executing the **SendAsync** in the **Fire And Forget** Task !

Which means the `server.Dispose()` will be called **Before** the completion of **SendAsync**.

**Try this two options :** 

**`client.SendAsync("aaa").GetAwaiter.GetResult();`** ( without Task.Run) and see what happens.

**Or**

**`await client.SendAsync("aaa");`**

Hope this helps.👍

**Comment by sancheolz** [comment]
@ShayanFiroozi Hi. The client can be on another machine and send messages independently from the server. This example simulates such a scenario. So waiting doesn't make sense.

**Comment by jchristn** [comment]
Moving this to discussion.  The right approach is to `await` the `client.SendAsync` so that it has to complete before the server is disposed.

**Reply by sancheolz** [reply]
Adding await will fix this test, but it won't fix the real-world scenario where the client is on a different host. I think the issue is that the server is still processing the incoming packet, and at that moment, the server's Dispose is being called. I expect the server's Dispose to not throw an exception, regardless of when it is called - either while the incoming message is being processed or when there are no messages. This is one of the rules for writing Dispose - it should not throw exceptions.
